azurelinuxagent/ga/update.py

```diff
@@ -1288,6 +1294,27 @@ def _send_heartbeat_telemetry(self, protocol):
             self._heartbeat_update_goal_state_error_count = 0
             self._last_telemetry_heartbeat = datetime.utcnow()
 
+    def _check_agent_memory_usage(self):
+        """
+        This checks the agent current memory usage and safely exit the process if agent reaches the memory limit
+        """
+        try:
+            if conf.get_enable_agent_memory_usage_check() and self._extensions_summary.converged:
+                if self._last_check_memory_usage == datetime.min or datetime.utcnow() >= (self._last_check_memory_usage + UpdateHandler.CHECK_MEMORY_USAGE_PERIOD):
+                    self._last_check_memory_usage = datetime.utcnow()
+                    CGroupConfigurator.get_instance().check_agent_memory_usage()
+        except CGroupsException as exception:
+            msg = "Check on agent memory usage:\n{0}".format(ustr(exception))
+            logger.info(msg)
+            add_event(AGENT_NAME, op=WALAEventOperation.CGroupsInfo, is_success=True, message=msg)
+            raise ExitException("Agent {0} is reached memory limit -- exiting".format(CURRENT_AGENT))
```

Comment:
This ExitException and rest of the logic of sys.exit already handled in main thread for safe exit.